src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs

```diff
@@ -24,18 +25,34 @@ public RemoteDependencyWalker(RemoteWalkContext context)
             _context = context;
         }
 
-        public Task<GraphNode<RemoteResolveResult>> WalkAsync(LibraryRange library, NuGetFramework framework, string runtimeIdentifier, RuntimeGraph runtimeGraph, bool recursive)
+        public async Task<GraphNode<RemoteResolveResult>> WalkAsync(LibraryRange library, NuGetFramework framework, string runtimeIdentifier, RuntimeGraph runtimeGraph, bool recursive)
         {
-            return CreateGraphNode(library, framework, runtimeIdentifier, runtimeGraph, _ => recursive ? DependencyResult.Acceptable : DependencyResult.Eclipsed, outerEdge: null);
+            var transitiveCentralPackageVersions = new TransitiveCentralPackageVersions();
+            var rootNode = await CreateGraphNode(
+                libraryRange: library,
+                framework: framework,
+                runtimeName: runtimeIdentifier,
+                runtimeGraph: runtimeGraph,
+                predicate: _ => (recursive ? DependencyResult.Acceptable : DependencyResult.Eclipsed, null),
+                outerEdge: null,
+                transitiveCentralPackageVersions: transitiveCentralPackageVersions);
+
+            while (transitiveCentralPackageVersions.TryTake(out LibraryDependency centralPackageversionDependecy))
+            {
+                await AddTransitiveCentralPackageVersionNodesAsync(rootNode, centralPackageversionDependecy, framework, runtimeIdentifier, runtimeGraph, transitiveCentralPackageVersions);
+            }
```

Comment:
Here's an alternative that doesn't need the `TransitiveCentralPackageVersions` inner class.

What do you think?

If performance is a concern, the innerNodes foreach loop can check that the node is a project, not a package, before pushing it onto the stack.

```suggestion
            var nodesToCheck = new Stack<GraphNode<RemoteResolveResult>>();
            nodesToCheck.Push(rootNode);

            while (nodesToCheck.Count > 0)
            {
                var node = nodesToCheck.Pop();

                foreach (var dependency in node.Item.Data.Dependencies)
                {
                    if (dependency.VersionCentrallyManaged && dependency.ReferenceType == LibraryDependencyReferenceType.Transitve)
                    {
                        await AddTransitiveCentralPackageVersionNodesAsync(rootNode, dependency, framework, runtimeIdentifier, runtimeGraph);
                    }
                }

                foreach (var innerNode in node.InnerNodes)
                {
                    nodesToCheck.Push(innerNode);
                }
            }
```

Reply:
> Here's an alternative that doesn't need the `TransitiveCentralPackageVersions` inner class.
> 
> What do you think?
> 
> If performance is a concern, the innerNodes foreach loop can check that the node is a project, not a package, before pushing it onto the stack.

Couple of observations regarding the suggestion:
1. There is not any need to loop through the entire dependency list in node.Item.Data.Dependencies.
2. There should not be duplicates in the nodes. With the above suggestion are going to be. 

Observation about the current implementation:
The current `TransitiveCentralPackageVersions` can be thought as concurrent queue that does not allow duplicate enqueues. The items are added during the graph traversal with no need to traverse again the inner nodes. 
What do you think is incorrect about the current approach?

Reply:
> 1. There is not any need to loop through the entire dependency list in node.Item.Data.Dependencies.

There is if you don't keep a collection of used centrally managed transitive dependencies, which was the entire point of my suggestion. At least, that's the only place I could find ReferenceType to check if it's value is `None` or `Transitive` (find which ones were used, ignore the ones not used).

> 2. There should not be duplicates in the nodes. With the above suggestion are going to be.

In what scenario will there be duplicates? I thought the `PackageVersion` items are de-duplicated before restore, just as PackageReference and PackageDownload items are, so by the time it gets to this method there are no duplicate items. Is it because if a CPVM project has a project reference to another CPVM project, that child project is going to have its own items, which could be duplicates?

This makes me curious if there's a bug in both of our implementations: Should centrally managed versions of project references even be in the current project's restore graph? In particular, if the other project uses a different `Directory.Packages.props` file with different packages and versions defined, would we even want those other proejct's `PackageVersion` items from appearing, and therefore changing, the current project's restore graph? I haven't yet had a look at your tests, but you can answer this question quickly: Do you have a test where two projects use different `Directory.Packages.props` files that define different versions for packages from each other, and one of the projects has a `ProjectReference` to the other?  What's the expected behaviour, particularly in edge cases when the project being restored does not have a `PackageVersion` defined in its `DPp` file, but the referenced project does, which is different to the version defined in the package's dependencies list? What about when only one of the two projects are CPVM opt-in?

> What do you think is incorrect about the current approach?

Nothing incorrect (Except maybe the case I just described above, where project references use different CPVM files), but I found it difficult to understand. Before debugging to make my proposal I didn't know anything at all about this code, and that's where most of my difficulty understanding comes from. But when I see someone use `Concurrent` classes, to do a good job reviewing, I have to ask myself, why is a concurrent data structure used? So, I spent time to find out that dependencies are processed in parallel. Ok, since it's in parallel and modifying a single object's instance, it's good that concurrent data structures are used, but are they used correctly? Is there a potential parallel timing bug? Some people are big fans of lock-free concurrent algorithms, because it's easier to understand and less prone to bugs, because it's hard to do parallel/concurrent programming right. Even with all the APIs and data structures, highly skilled developers still make mistakes. So, even if you've done it right, I feel obligated to do a more in-depth review to make sure I think it's correct.

Additionally, not only did you use the `ConcurrentQueue` data structure in the implementation, you also described the change as "can be thought as concurrent queue". In my mind, queues have the characteristic of being a first-in, first-out collection, and is typically used in consumer-producer patterns. So now I'm asking myself why does the order matter? Why are things being removed in the queue while other tasks/threads are still adding things to the queue? As far as I can tell, that's not the case. It's just getting the set of items that were "promoted" from not used to a transitive dependency and the semantics are "write-only until graph walking is complete, then read-only". You could have achieved the same thing using only a `ConcurrentDictionary`, since there is no `ConcurrentSet`, and then iterating the dictionary's keys.

Maybe I'm overthinking it, but this is how I do code reviews, and why it takes me a long time. Something else I try to consider is if someone else is going to modify this code in 12 months from now. Will the code be easy to understand and modify?

My high level description of your implementation is:
* Recursively walk the dependencies graph
  * When a project is evaluated, add fake dependencies for all the package versions that do not already have a direct dependency.
  * When a "conflict" is found, and it's for one of these fake dependencies, mark the dependency as used and add it to a collection of used transitive dependencies.
* When graph walking is finished, get the used dependencies from the collection, and add them as real dependencies in the graph.

My proposal's high level description is like this:
* Recursively walk the dependencies graph
  * When a project is evaluated, add fake dependencies for all the package versions that do not already have a direct dependency.
  * When a "conflict" is found, and it's for one of these fake dependencies, mark the dependency as used.
* When graph walking is finished, find the used fake dependencies and add them as real dependencies.

When writing it out like this, it gave me a different idea:
* Don't add fake dependencies, and do a graph walk to download the full dependency graph, just like non-CPVM.
* Enumerate the graph to find all dependencies, and when the package ID matches a pinned package, add a top level dependency for it.

If the scenario at the top of this ridiculously long comment is really a bug (project reference with different CPVM files), not only will this last idea be easier for people to understand, and possibly have simpler code as a result, it won't be susceptible to the bug.

Reply:
I will try to address the points in the comment above. 
1. I do have a set of E2E tests used to validate scenarios like the ones mentioned. In case the Project1 ->Project2. Both use CPVM and different versions in the two Directory.Packages.props the main project dependencies wins. 
2. The two logic sequences above are similar. In the current implementation the graph is not traversed multiple times. The dependencies that need to be added are marked at the moment the graph nodes are created using this new type. In the other proposal the new type is not needed, it is eventually replaced with a Concurrent Collection but reiterate multiple times through the graph.
3. It was a comment about duplication. Imagine 
Project1 -> PackageA -> PackageD -> PackageB
                                    -> PackageC -> PackageB
Central PackageB
In this case the PackageB should be only marked once to be added to the graph. 
4. I am not sure if I understated well the comments regarding the concurrent need for the `TransitiveCentralPackageVersions` - the adds are done from different threads. 
5. There was a comment regarding the ordering. It could be very well a ConcurrentBag ( that is implemented on top of ConcurrentQueue). I mentioned it only trying to explain more the logic. The `TransitiveCentralPackageVersions` has only two methods (TryAdd, TryTake) none of them implying an order. 
6. I understand that the new changes are difficult to understand. I will add more comments to make it more clear.

Reply:
I just wanted to add a specific emphasis to one of @zivkan's comments that I really agree with: 

> Don't add fake dependencies, and do a graph walk to download the full dependency graph, just like non-CPVM.

fwiw, I think the fake dependency is confusing. 
It's a separate concept that's hard to follow right now. 
The fact that the library range can now represent a transitive dependency is confusing.  
The central dependencies could be a only one object pasted to the walker when appropriate.